src/PluginClientBuilder.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpClient;
+use Http\Client\HttpAsyncClient;
+use Psr\Http\Client\ClientInterface;
+
+/**
+ * Build an instance of a PluginClient with a dynamic list of plugins.
+ *
+ * This builder is immutable, each "build" methods will instead create a new
+ * instance of itself with or without the affected plugin, destroying a
+ * potentially http client built in the new instance.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ */
+final class PluginClientBuilder
+{
+    /** @var Plugin[] */
+    private $plugins = [];
+
+    /** @var ClientInterface | HttpClient | HttpAsyncClient */
+    private $client;
+
+    /** @var ?PluginClient */
+    private $pluginClient;
```

Comment:
i agree with mark, would not cache the resulting client. we can create multiple instances of the same client with the builder, or even add more plugins and create another client with the additional plugins.